cpp/src/gandiva/lower_value_used_cache.h

```diff
@@ -0,0 +1,103 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <list>
+#include <set>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/optional.h"
+#include "gandiva/base_cache.h"
+
+// modified cache to support evict policy of lower value used.
+namespace gandiva {
+// a cache which evicts the lower value used item when it is full
+template <class Key, class Value>
+class LowerValueUsedCache : public BaseCache<Key, Value> {
```

Comment:
Just evicting based on the build times seems would be problematic since it might just keep only infrequently used large expressions. You should use a combination of build time and some aging policy.
Also can you change the name. It sounds confusing.

Reply:
I agree with you!
What I want to discuss here is what would be the best trigger to remove an element according to the build time policy or the recency policy. Do you have any opinion about it?

Reply:
Something like https://en.wikipedia.org/wiki/Cache_replacement_policies#LFU_with_dynamic_aging_(LFUDA)